lib/IRGen/IRGenModule.cpp

```diff
@@ -1185,63 +1175,186 @@ static bool isFirstObjectFileInModule(IRGenModule &IGM) {
+      break;
+    }
+    case AutolinkKind::SwiftAutoLinkExtract: {
+      // Merge the entries into null-separated string.
+      llvm::SmallString<64> EntriesString;
+      for (auto EntryNode : Entries) {
+        const llvm::MDNode *MD = cast<llvm::MDNode>(EntryNode);
+        for (auto &Entry : MD->operands()) {
+          const llvm::MDString *MS = cast<llvm::MDString>(Entry);
+          EntriesString += MS->getString();
+          EntriesString += '\0';
+        }
+      }
+      auto EntriesConstant = llvm::ConstantDataArray::getString(
+          IGM.getLLVMContext(), EntriesString, /*AddNull=*/false);
+      // Mark the swift1_autolink_entries section with the SHF_EXCLUDE attribute
+      // to get the linker to drop it in the final linked binary.
+      // LLVM doesn't provide an interface to specify section attributs in the
+      // IR so we pass the attribute with inline assembly.
+      if (IGM.TargetInfo.OutputObjectFormat == llvm::Triple::ELF)
+        IGM.Module.appendModuleInlineAsm(".section .swift1_autolink_entries,"
+                                         "\"0x80000000\"");
+      auto var =
+          new llvm::GlobalVariable(*IGM.getModule(), EntriesConstant->getType(),
+                                   true, llvm::GlobalValue::PrivateLinkage,
+                                   EntriesConstant, "_swift1_autolink_entries");
+      var->setSection(".swift1_autolink_entries");
+      var->setAlignment(llvm::MaybeAlign(IGM.getPointerAlignment().getValue()));
+
+      disableAddressSanitizer(IGM, var);
+      IGM.addUsedGlobal(var);
+      break;
+    }
+    }
+  }
+
+  static AutolinkKind create(const SwiftTargetInfo &TargetInfo,
+                             llvm::Triple Triple,
+                             IRGenLLVMLTOKind LLVMLTOKind) {
+    // When performing LTO, we always use lld that supports auto linking
+    // mechanism with ELF. So embed dependent libraries names in
+    // "llvm.dependent-libraries" instead of "llvm.linker.options".
+    if (TargetInfo.OutputObjectFormat == llvm::Triple::ELF &&
+        LLVMLTOKind != IRGenLLVMLTOKind::None) {
+      return AutolinkKind::LLVMDependentLibraries;
+    } else if (doesTargetAutolinkUsingAutolinkExtract(TargetInfo, Triple)) {
+      return AutolinkKind::SwiftAutoLinkExtract;
+    } else {
+      return AutolinkKind::LLVMLinkerOptions;
+    }
+  }
+};
+} // namespace
+
 void IRGenModule::emitAutolinkInfo() {
+
+  auto Autolink =
+      AutolinkKind::create(TargetInfo, Triple, IRGen.Opts.LLVMLTOKind);
+
+  StringRef AutolinkSectionName = Autolink.getSectionNameMetadata();
+
+  auto *Metadata = Module.getOrInsertNamedMetadata(AutolinkSectionName);
+  llvm::SmallVector<llvm::MDNode *, 4> Entries;
+
   // Collect the linker options already in the module (from ClangCodeGen).
-  // FIXME: This constant should be vended by LLVM somewhere.
-  auto *Metadata = Module.getOrInsertNamedMetadata("llvm.linker.options");
   for (llvm::MDNode *LinkOption : Metadata->operands())
```

Comment:
Don't think you need to make this change, but a one liner you can use for this is:

```
llvm::copy(Metadata->operands(), std::back_inserter(Entries));
```